### ISSUE TITLE
Native crash when using LiveQuery

### ISSUE THREAD

**Issue by SzyQ** [couchbase/couchbase-lite-android#949]
It looks like there is a memory leak in the cbforest lib, which triggers the crash from assertion error.
Line 115: https://github.com/couchbaselabs/cbforest/blob/master/C/c4Impl.hh
Logs:
```
> 07-28 14:10:31.899 D/CouchbaseDBImpl(12334): Database opened: reference_data_resources
07-28 14:10:31.919 D/Query (12334): com.couchbase.lite.LiveQuery@f98c8a5: start() called
07-28 14:10:31.919 V/Query (12334): com.couchbase.lite.LiveQuery@f98c8a5: start() is calling update()
07-28 14:10:31.919 V/Query (12334): com.couchbase.lite.LiveQuery@f98c8a5: update() called.
07-28 14:10:31.919 D/Query (12334): com.couchbase.lite.LiveQuery@f98c8a5: update() created queryFuture: java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@4a3a42b
07-28 14:10:31.923 E/cbforest(12334): invalid compaction mode (-22)
07-28 14:10:31.923 E/cbforest(12334): Assertion failed: _refCount == 0 (../vendor/cbforest/C/c4Impl.hh:115, in virtual c4Internal::RefCounted::~RefCounted() [SELF = c4View])
07-28 14:10:31.923 F/libc (12334): Fatal signal 6 (SIGABRT), code -6 in tid 13472 (CBLManagerWorkE)
07-28 14:10:32.024 F/DEBUG (20900): *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
07-28 14:10:32.024 F/DEBUG (20900): Build fingerprint: 'google/hammerhead/hammerhead:6.0.1/MOB30P/2960889:user/release-keys'
07-28 14:10:32.024 F/DEBUG (20900): Revision: '11'
07-28 14:10:32.024 F/DEBUG (20900): ABI: 'arm'
07-28 14:10:32.025 F/DEBUG (20900): pid: 12334, tid: 13472, name: CBLManagerWorkE >>> com.example.debug <<<
07-28 14:10:32.025 F/DEBUG (20900): signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
07-28 14:10:32.037 F/DEBUG (20900): r0 00000000 r1 000034a0 r2 00000006 r3 9e07f978
07-28 14:10:32.037 F/DEBUG (20900): r4 9e07f980 r5 9e07f930 r6 00000000 r7 0000010c
07-28 14:10:32.037 F/DEBUG (20900): r8 b3f71050 r9 9e07f080 sl aee6f600 fp aee67300
07-28 14:10:32.038 F/DEBUG (20900): ip 00000006 sp 9e07eea0 lr b6c6cb61 pc b6c6ef50 cpsr 400f0010
07-28 14:10:32.048 F/DEBUG (20900): 
07-28 14:10:32.048 F/DEBUG (20900): backtrace:
07-28 14:10:32.048 F/DEBUG (20900): #00 pc 00041f50 /system/lib/libc.so (tgkill+12)
07-28 14:10:32.049 F/DEBUG (20900): #1 pc 0003fb5d /system/lib/libc.so (pthread_kill+32)
07-28 14:10:32.049 F/DEBUG (20900): #2 pc 0001c30f /system/lib/libc.so (raise+10)
07-28 14:10:32.049 F/DEBUG (20900): #3 pc 000194c1 /system/lib/libc.so (__libc_android_abort+34)
07-28 14:10:32.049 F/DEBUG (20900): #4 pc 000174ac /system/lib/libc.so (abort+4)
07-28 14:10:32.049 F/DEBUG (20900): #5 pc 001def57 /data/app/com.example.debug-1/lib/arm/libCouchbaseLiteJavaForestDB.so (_ZN9__gnu_cxx27__verbose_terminate_handlerEv+226)
07-28 14:10:32.049 F/DEBUG (20900): #6 pc 001afe95 /data/app/com.example.debug-1/lib/arm/libCouchbaseLiteJavaForestDB.so (_ZN10__cxxabiv111__terminateEPFvvE+4)
07-28 14:10:32.049 F/DEBUG (20900): #7 pc 001aff09 /data/app/com.example.debug-1/lib/arm/libCouchbaseLiteJavaForestDB.so (_ZSt9terminatev+8)
07-28 14:10:32.049 F/DEBUG (20900): #8 pc 000bda41 /data/app/com.example.debug-1/lib/arm/libCouchbaseLiteJavaForestDB.so
```

**Comment by hideki** [comment]
Hi @SzyQ,

Thank you for bug report and crash log.

If you have time, could you please try `1.3.0-45` build? This has fix of concurrent query issue with ForestDB. 
Note: fix for `1.3.0-45` will not be in `1.3.0`, this should be in next version.
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.0-45/
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android-forestdb/1.3.0-45/

Thanks!

**Comment by hideki** [comment]
ref #815

**Comment by SzyQ** [comment]
Hi,

With version 1.3.0-45 I get the same error:

> 07-29 08:06:08.589 E/cbforest(23715): invalid compaction mode (-22)
07-29 08:06:08.589 E/cbforest(23715): Assertion failed: _refCount == 0 (../vendor/cbforest/C/c4Impl.hh:115, in virtual c4Internal::RefCounted<c4View>::~RefCounted() [SELF = c4View])

**Comment by hideki** [comment]
@SzyQ, Thank you for your inputs. I will look into this issue

**Comment by SzyQ** [comment]
Appreciate it, thanks

**Comment by SzyQ** [comment]
Hi @hideki  Any progress on that?

**Comment by hideki** [comment]
sorry. no progress.

**Comment by sethrosetter** [comment]
@pasin Can you follow up with me on adding a regression / unit test?

**Comment by hideki** [comment]
Hi @SzyQ,
Sorry for not taking care this issue soon. I am currently looking into this issue. Our test cases covers LiveQuery with forestdb storage type. I can not see the crash in forestdb with using build from master branch. http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.1-44/  Do you still face this issue? If so, can you provide details? How many LiveQueries running concurrently? Each query's parameter? and Each view's map function?
Thanks!

**Comment by hideki** [comment]
If you are still facing issue, could you please try `1.3.1-56` build?
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.1-56/

**Comment by SzyQ** [comment]
Hi @hideki, Sorry for late response. I've build the app with version 1.3.1-56 and the issue no longer occurs. Will it be included in v1.3.1?

**Comment by hideki** [comment]
Hi @SzyQ,
Thank you for verifying the fix.
This fix should be in next minor release. But I don't know next miner release version. It could be v1.4.0 or 1.5.0 as Couchbase Lite might skip v1.4.0. 
Thanks!
